Improve unchecked type arg warnings

som-snytt opened the pull request:
Moved text around for readability, tweak is at `isWarnableTypeArg` but is a kludge.

Library clean-up precedes the warning change, but test updates are committed with it.

Fixes https://github.com/scala/bug/issues/12408

som-snytt committed:
Refactor checkable and use set

Textual refactor for readability. Prefer local defs.

Previous optimization used ListBuffer to collect type args,
but since a Set is required, use one directly.

som-snytt committed:
Update lib for uncheckable type args

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Checkable.scala:
```diff
@@ -179,20 +162,48 @@ trait Checkable {
+        //import util._
+        def isUnwarnableTypeArgSymbol(sym: Symbol) = {
+          @`inline` def exemption = sym.owner == tp.typeSymbol || inPattern && (sym.owner.name == nme.unapply || sym.owner.name == nme.unapplySeq)
+          sym.isTypeParameter && exemption ||     // dummy
```
OK, so `dummy` refers to https://github.com/scala/scala/blob/37cf33ac78aee3ce457b638de62106c4c3be4dd4/src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala#L394, and `exemption` is trying to capture this more precisely?

Maybe we can find a better secret handshake between the two methods?

In this example

```
class Renderer[T]
class Test[A](a: Any) {
  def f2[B] = a match { case _: Renderer[B] => }
  def g = a match { case _: Renderer[A] => }
}
```

I don't get why `sym.isTypeParameter` is true for the `A` in `case _: Renderer[A]`, but it's false for the `B` in `case  _: Renderer[B]`. I guess it's because I still don't get type parameter skolems.

som-snytt replied:
I looked a bit at what might differentiate syms here vs passing flags down the stack for more context, or add attachment somewhere. I pushed an update that is still WIP. But I don't have a complete picture yet.

som-snytt commented:
The second error was out of disk space.
```
[error] Caused by: sbt.ForkMain$ForkError: java.io.FileNotFoundException: /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/pos/spec-partially-pos.log (No space left on device)
```
IIRC it creates log files as placeholder even if there is no output? Normally if you kill partest, you'll see empty leftover logs that weren't cleaned up. That's not necessarily a great system.

som-snytt commented:
/rebuild

lrytz committed:
More correcter type patterns in synthetic code

Synthetic `case _: Foo[_]` patterns were generated with a weird
`TypeRef(pre, fooSymbol, args)` where the `args` are Foo's type
parameters.

Now we generate something closer to what the type checker would do.

som-snytt commented:
Something like that. :)

lrytz committed:
fix more uncheckable type args in the library

lrytz commented:
@som-snytt how does it look to you now?

dwijnand commented on test/files/run/t12405.check:
```diff
@@ -24,7 +24,7 @@ package <empty> {
   case <synthetic> val x1: Any = x$1;
   case5(){
-    if (x1.isInstanceOf[C[A]])
+    if (x1.isInstanceOf[C[$this._]])
```
Looks wrong, no?

lrytz replied:
I think it's right, the synthetic `equals` method used to match `case _: C[A]` where `C[A]` was a weird ill-formed TypeRef in which the argument symbol `A` is class C's type parameter symbol. Now the synthetic `equals` method becomes `case _: C[_]`, where `_` references new AbstractTypeSymbol (this is the change in SyntheticMethods), same as in a hand-written match.

Maybe `_`'s owner is wrong though (`$this`), not the same as what the typer would generate. But I don't think that matters (EDIT: as this symbol is never entered into any scope).

som-snytt commented:
Looks more gooder to me. Makes me feel like a "dummy".

lrytz commented:
It certainly made me feel like a dummy for more than two days.

lrytz commented:
@SethTisue would you mind running this through a community build? It uncovers a lot of missing unchecked warnings, see the diff for what's needed in the standard library.

I should also compare this PR to Scala 3's unchecked warnings.

som-snytt commented:
Let me know if you'd like me to clean up my commits, especially the reverted dummy bit. Or you can feel free to squash or what have you.

The warning tweak was inspired by Scala 3, I don't remember how much comparing I did.

dwijnand commented on src/compiler/scala/tools/nsc/typechecker/Checkable.scala:
```diff
@@ -39,7 +37,7 @@ import scala.tools.nsc.Reporting.WarningCategory
  *  There are four possibilities to consider:
  *     [P1] X will always conform to P
- *     [P2] x will never conform to P
+ *     [P2] X will never conform to P
```
I think this isn't a typo: I think it's from `x ∉ P              // (x ∈ $X) ⇒ (x ∉ $P)`

som-snytt replied:
I think you're right, but doesn't it need a different verb? I'll fix that with any other clean up, if desired.

som-snytt committed:
Tweak message if multiple unchecked args.

som-snytt committed:
Tweaks for readability, provisional test

som-snytt commented:
@dwijnand is this warning intended? https://github.com/scala/scala/pull/9672/files#diff-d1d7b8beefa8bf0159896c33a250f3e8b2307bc673e6672e0082d8a7b9654ae2R163

where dotty does not warn. It's extractor of Any.

som-snytt commented:
The neg test, extracted from `run/patmat-exprs.scala`, shows `nowarn` getting copied to synthetic methods; commenting out the pattern match reduces the warnings; so the warnings depend on typechecking, which is why we need a tool to compare trees before/after an edit. Not sure if nowarn or synthetic is to blame.

som-snytt commented:
Asked at https://github.com/lampepfl/dotty/issues/13010

dwijnand commented:
Which one?  Do you mean `a => a match { case x: Expr[T @unchecked] => x.optimizeWith(f) }`?

som-snytt commented:
@dwijnand sorry the UI doesn't like the link, which I improved. ^  You might have to make one of those instructional videos so we all have canonical expectations.

dwijnand commented:
> @dwijnand is this warning intended? [#9672 (files)](https://github.com/scala/scala/pull/9672/files#diff-d1d7b8beefa8bf0159896c33a250f3e8b2307bc673e6672e0082d8a7b9654ae2R163)
> 
> that is, [https://github.com/scala/scala/blob/v2.13.6/test/files/run/patmat-behavior.check#L163](https://github.com/scala/scala/blob/v2.13.6/test/files/run/patmat-behavior.check?rgh-link-date=2021-07-02T21%3A13%3A56Z#L163)
> warning on [https://github.com/scala/scala/blob/v2.13.6/test/files/run/patmat-behavior.scala#L68](https://github.com/scala/scala/blob/v2.13.6/test/files/run/patmat-behavior.scala?rgh-link-date=2021-07-02T21%3A13%3A56Z#L68)
> 
> where dotty does not warn, case class scrutinee and random extractors taking Any.

I would think so?  Can you see a reasonable explanation for why that scenario should switch off the lights?  I think we should open a reproducer in the Scala 3 repo (where I might end up looking to fix it).

lrytz commented:
> The neg test, extracted from `run/patmat-exprs.scala`, shows `nowarn` getting copied to synthetic methods; commenting out the pattern match reduces the warnings; so the warnings depend on typechecking, which is why we need a tool to compare trees before/after an edit. Not sure if nowarn or synthetic is to blame.

I don't follow what you mean here; removing the `@nowarn` annotations doesn't seem to change anything (besides removing "@nowarn annotation does not suppress any warnings")

SethTisue commented:
https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/2980/ queued

som-snytt commented:
@lrytz it warns only for `case class`, and emits both param unused and nowarn suppresses nothing. I'll take a second look at some point.

SethTisue commented:
Community build run finished, and the one failure is a known intermittent failure. We still don't have https://github.com/scala/community-build/issues/1270 so good luck groveling through the logs :-)

som-snytt commented:
I tried to follow it yesterday but only saw lots of other warnings and infrastructure output.

lrytz commented:
Did some manual log diff fiddling
 - https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/2980 is the build for this PR
 - 2943 / 2944 are the builds for the last Scala bump, according to https://github.com/scala/community-build/pull/1439

```
curl https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/2980/consoleText > log-pr.txt
curl https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/2943/consoleText > log-base-1.txt
curl https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/2944/consoleText > log-base-2.txt

grep -h 'is unchecked\|cannot be checked' log-pr.txt | sort | sed 's;/home.*[0-9a-f\]\{40\};;' > unchecked-pr.txt
grep -h 'is unchecked\|cannot be checked' log-base-1.txt log-base-2.txt | sort | sed 's;/home.*[0-9a-f\]\{40\};;' > unchecked-base.txt
```

https://gist.github.com/lrytz/55b13fdbe49fba76f9fcd454e754e91c

A lot of new warnings in akka-http, akka-more, akka-stream, akka and twitter-util. A few in other projects.

lrytz commented:
Looking at a few examples in akka, some of the warnings showed up when cross-building to Scala 3 and were recently annotated (https://github.com/akka/akka/commit/41d5f1571b9254e229989e9cc6618089e46625dc), the community build reference doesn't have this change yet.

SethTisue commented:
@lrytz @dwijnand final review/merge?

som-snytt commented:
This is the director's cut, where Lukas is the director. From now on, instead of "IDK ask Lukas", it should be, "ask The Director."

I would have to refresh my memory about my two previous questions, which did not bother anyone. Probably looking at the community build induced a traumatic break.

I have my usual reservations about any change that achieves parity with dotty, so-called parody, because it reduces the incentive to upgrade. OTOH, perhaps it also incentivizes maintenance subscriptions for Lightbend, so it cuts both ways.

lrytz commented:
Scala 3 still issues many more warnings than Scala 2 with this PR, e.g.,

```
-- Warning: ../src/library/scala/collection/immutable/Set.scala:101:11 ---------
101 |      case _: SortedSet[E]         => (newBuilder[E] ++= it).result()
    |           ^
    |the type test for scala.collection.immutable.SortedSet[E] cannot be checked at runtime

```

for https://github.com/scala/scala/blob/161380ece6325e596bf20d0e887ac3c4818a47d0/src/library/scala/collection/immutable/Set.scala#L97-L105

I'll take a look who's right. In this example, it seems to me that `IterableOnce[E] & SortedSet[_]` implies `SortedSet[E]`, is it possible that Scala 2 knows/considers that but Scala 3 doesn't? (@dwijnand or @smarter might have a clue)

smarter commented:
>  it seems to me that IterableOnce[E] & SortedSet[_] implies SortedSet[E], is it possible that Scala 2 knows/considers that but Scala 3 doesn't?

No, Scala 3 is right here: IterableOnce is covariant whereas SortedSet is invariant, so if I have a SortedSet[Int], I can upcast it to IterableOnce[Any], but not SortedSet[Any].

som-snytt commented:
Next coffee break, I will ponder the comment about "element type widening issues", and also ponder why it sounds obvious when smarter says it.

lrytz commented:
Thanks @smarter. For example-oriented people like me:

```scala
scala> def t[T](x: IterableOnce[T], e: T) = x match {
     |   case s: Set[T] => s(e)
     |   case _ => false
     | }
def t[T](x: IterableOnce[T], e: T): Boolean

scala> t[Any](collection.immutable.BitSet(1), "hi")
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
```

Scala 3 warns at the type pattern.

lrytz reviewed (approved):
I compared the new warnings that this PR issues on the standard library to the ones from Scala 3. All new warnings from this PR are also issued by Scala 3, which is good validation.

The remaining warnings that are missing here but issued by Scala 3 seem to be due to https://github.com/scala/bug/issues/12439.

Let's book the progress.
